### PR TITLE
bring up-to-date with canonical-kubernetes

### PR DESCRIPTION
**Pull request by Cynerva**
Work for https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/70

This pulls in additional functionality from canonical-kubernetes (flannel, ingress, etc) while trying to maintain a low machine footprint.

This includes an updated README as well, [see here](https://github.com/juju-solutions/bundle-kubernetes-core/tree/gkk/converged).

@chuckbutler @mbruzek @wwwtyro @marcoceppi

### CODE REVIEW

**Comment by wwwtyro** [comment]
:+1: LGTM

**Comment by lazypower** [comment]
This looks fine to me. _disclaimer_ I haven't combed the README for consistency and accuracy, but the bundle itself looks good.

**Comment by lazypower** [comment]
@Cynerva - i know this is late in the game, but we did just rev the k8s charms to -4

would you mind bumping the charm revisions in here to reflect that?

File: README.md [hunk 1]
```diff
@@ -1,123 +1,404 @@
...
 [machine constraints documentation](https://jujucharms.com/docs/stable/charms-constraints)
-if you want more information about setting larger constraints.
+for other machine constraints that might be useful for the kubernetes-worker units.
+
 
 ### Scaling Etcd
 
 Etcd is used as a key-value store for the Kubernetes cluster. For reliability
-the cluster defaults to three instances in this bundle.
+the bundle defaults to three instances in this cluster.
```
Comment (mbruzek) [inline README.md]: The number of etcd units are 1 in this new bundle. We should change 'three' to 'one' here.

File: bundle.yaml [hunk 2]
```diff
@@ -1,21 +1,85 @@
...
-  kubernetes:
-    charm: cs:xenial/kubernetes-8
-    options:
-      version: v1.3.4
+machines:
+  0:
+  1:
+  2:
+  3:
+applications:
```
Comment (mbruzek) [inline bundle.yaml]: The key 'applications' is a v4 bundle convenation that is not yet widely supported. @chuckbutler had to change this key back to 'services' in the other bundle for it to work with our testing tools.

File: bundle.yaml [hunk 3]
```diff
@@ -1,21 +1,85 @@
...
+  "kubernetes-master":
+    charm: "cs:~containers/kubernetes-master-3"
```
Comment (mbruzek) [inline bundle.yaml]: We **just** incremented the version of the kubernetes-master charm last night. s/3/4/

File: bundle.yaml [hunk 4]
```diff
@@ -1,21 +1,85 @@
...
+  easyrsa:
+    charm: "cs:~containers/easyrsa-2"
+    num_units: 1
+    to:
+      - lxd:0
+    annotations:
+      "gui-x": "450"
+      "gui-y": "550"
+  "kubernetes-worker":
+    charm: "cs:~containers/kubernetes-worker-3"
```
Comment (mbruzek) [inline bundle.yaml]: We also published a new kubernetes-worker with fixes. Can you change this to 4 as well?

**Review by mbruzek — COMMENTED** [review]
Great work on this README and bundle! I have a few changes that would make this more accurate.

**Comment by Cynerva** [comment]
Updated:
- Bumped revisions for kubernetes-worker, kubernetes-master, and etcd
- Used `services:` instead of `applications:`
- Fixed etcd unit count in README

Have not tested with the bundle changes yet; will do so later today.

**Comment by Cynerva** [comment]
Updated:
- Moved etcd to machine 0 (shared with kubeapi-load-balancer)

Again, not tested yet.

**Comment by mbruzek** [comment]
Thanks for addressing my comments @Cynerva ! +1 to these changes.

**Pull request merged**